Improve handling of error types in dagger factory generation

### Discussed in https://github.com/ZacSweers/anvil/discussions/62

<div type='discussions-op-text'>

<sup>Originally posted by **bddckr** August 26, 2024</sup>
I have a class `A` that depends on class `B`:
```kotlin
class A @Inject constructor(private val b: B)
```

`B` is generated by my own code generator, which is also using KSP.

I'm getting the following error from the `kspDebugKotlin` task (this is an Android library module) when using KSP2:
> [ksp] Error type '&lt;ERROR TYPE: B&gt;' is not resolvable in the current round of processing. Check your imports or, if this is a generated type, ensure the tool that generates it has its outputs appropriately sources as inputs to the KSP task.

> &gt; A failure occurred while executing com.google.devtools.ksp.gradle.KspAAWorkerAction
>   &gt; KSP failed with exit code: PROCESSING_ERROR

When using KSP1, the error message is pretty much the same - it just mentions the file name as well, which helps when debugging:
> [ksp] /Users/bddckr/Projects/ksp-testing/some-module/src/main/kotlin/A.kt:5: Error type '&lt;ERROR TYPE: B&gt;' is not resolvable in the current round of processing. Check your imports or, if this is a generated type, ensure the tool that generates it has its outputs appropriately sources as inputs to the KSP task.

I disabled all the KSP processors/dependencies one-by-one to identify whether this is from dagger-ksp or Anvil. It's Anvil as far as I can see - this error isn't raised when I run with just my generator and dagger-ksp.

I've tried using `AnvilKspExtension` while adding a custom annotation to the class definition of `A`, so I can return the fully qualified name of that annotation via `AnvilKspExtension.supportedAnnotationTypes`, but it doesn't help: The error is raised _after_ my custom code generator. Without usage of `AnvilKspExtension`, the error is raised _before_ my code generator executes. (I've learned that by logging via the environment's `KSPLogger` in my generator.)

---

How do I get Anvil's symbol processor to "ignore" the missing type until my generator had a chance to create it in the current round?

Is the usage of `AnvilKspExtension` required? If so:
* What am I doing wrong?
* Is there a way to use `AnvilKspExtension` without requiring the use of an annotation? As you can see, in my use case there's no annotation. (My code generator looks up files and class declarations based on some naming convention - no annotation-based lookup is done at all.)

Side-note: This is all working fine with a custom generator and upstream Anvil v2.5.0-beta11. It's only now that I'm moving everything to KSP that I'm having an issue with my use case.

It looks like KSP expects the Anvil processor to not fail, but return the erroring symbols to defer to the next round. At least that's my understanding from [this](https://github.com/google/ksp/issues/1043#issuecomment-1187768871) and the [KSP docs on multi-round processing](https://kotlinlang.org/docs/ksp-multi-round.html).</div>

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -3,6 +3,8 @@
 **Unreleased**
 --------------
 
+- **Enhancement**: Better handle error types and round deferral for generated types.
+
 0.3.0
 -----
 

--- a/compiler/src/main/java/com/squareup/anvil/compiler/Utils.kt
+++ b/compiler/src/main/java/com/squareup/anvil/compiler/Utils.kt
@@ -221,3 +221,10 @@ internal val ClassName.fqName: FqName
 
 internal inline fun <T, C : Collection<T>, O> C.ifNotEmpty(body: (C) -> O?): O? =
   if (isNotEmpty()) body(this) else null
+
+internal infix fun <E> MutableList<E>.addWithoutCopy(e: E) = apply {
+  add(e)
+}
+internal infix fun <E> MutableList<E>.addAllWithoutCopy(e: Iterable<E>) = apply {
+  addAll(e)
+}

--- a/compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/AssistedFactoryCodeGen.kt
+++ b/compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/AssistedFactoryCodeGen.kt
@@ -34,6 +34,7 @@ import com.squareup.anvil.compiler.codegen.dagger.AssistedFactoryCodeGen.KspGene
 import com.squareup.anvil.compiler.codegen.ksp.AnvilSymbolProcessor
 import com.squareup.anvil.compiler.codegen.ksp.AnvilSymbolProcessorProvider
 import com.squareup.anvil.compiler.codegen.ksp.KspAnvilException
+import com.squareup.anvil.compiler.codegen.ksp.KspErrorTypeException
 import com.squareup.anvil.compiler.codegen.ksp.contextualToTypeName
 import com.squareup.anvil.compiler.codegen.ksp.isAnnotationPresent
 import com.squareup.anvil.compiler.codegen.ksp.isInterface
@@ -86,20 +87,26 @@ internal object AssistedFactoryCodeGen : AnvilApplicabilityChecker {
     class Provider : AnvilSymbolProcessorProvider(AssistedFactoryCodeGen, ::KspGenerator)
 
     override fun processChecked(resolver: Resolver): List<KSAnnotated> {
+      val deferred = mutableListOf<KSAnnotated>()
       resolver.getSymbolsWithAnnotation(assistedFactoryFqName.asString())
         .filterIsInstance<KSClassDeclaration>()
         .forEach { clazz ->
           generateFactoryClass(clazz)
-            .writeTo(env.codeGenerator, aggregating = false, listOf(clazz.containingFile!!))
+            ?.writeTo(env.codeGenerator, aggregating = false, listOf(clazz.containingFile!!))
+            ?: deferred.add(clazz)
         }
-      return emptyList()
+      return deferred
     }
 
     private fun generateFactoryClass(
       clazz: KSClassDeclaration,
-    ): FileSpec {
+    ): FileSpec? {
       val typeParameterResolver = clazz.typeParameters.toTypeParameterResolver()
-      val function = clazz.requireSingleAbstractFunction(typeParameterResolver)
+      val function = try {
+        clazz.requireSingleAbstractFunction(typeParameterResolver)
+      } catch (e: KspErrorTypeException) {
+        return null
+      }
 
       val returnType = try {
         function.returnType
@@ -218,7 +225,11 @@ internal object AssistedFactoryCodeGen : AnvilApplicabilityChecker {
       // the version from the file we're parsing ensures the correct variance of the referenced types.
       // TODO can't use getAllFunctions() yet due to https://github.com/google/ksp/issues/1619
       val assistedFunctions = sequenceOf(this)
-        .plus(getAllSuperTypes().mapNotNull { it.resolveKSClassDeclaration() })
+        .plus(
+          getAllSuperTypes()
+            .onEach { if (it.isError) throw KspErrorTypeException() }
+            .mapNotNull { it.resolveKSClassDeclaration() },
+        )
         .distinctBy { it.qualifiedName?.asString() }
         .flatMap { clazz ->
           clazz.getDeclaredFunctions()
@@ -229,6 +240,7 @@ internal object AssistedFactoryCodeGen : AnvilApplicabilityChecker {
         }
         .distinctBy { it.simpleName.asString() }
         .map {
+          if (it.returnType?.resolve()?.isError == true) throw KspErrorTypeException()
           it.asMemberOf(implementingType)
             .toAssistedFactoryFunction(it, typeParameterResolver)
         }

--- a/compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/AssistedInjectCodeGen.kt
+++ b/compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/AssistedInjectCodeGen.kt
@@ -65,6 +65,7 @@ internal object AssistedInjectCodeGen : AnvilApplicabilityChecker {
     class Provider : AnvilSymbolProcessorProvider(AssistedInjectCodeGen, ::KspGenerator)
 
     override fun processChecked(resolver: Resolver): List<KSAnnotated> {
+      val deferred = mutableListOf<KSAnnotated>()
       resolver.injectConstructors()
         .forEach { (clazz, constructor) ->
           if (!constructor.isAnnotationPresent<AssistedInject>()) {
@@ -75,19 +76,20 @@ internal object AssistedInjectCodeGen : AnvilApplicabilityChecker {
             clazz = clazz,
             constructor = constructor,
           )
-            .writeTo(env.codeGenerator, aggregating = false, listOf(constructor.containingFile!!))
+            ?.writeTo(env.codeGenerator, aggregating = false, listOf(constructor.containingFile!!))
+            ?: deferred.add(clazz)
         }
-      return emptyList()
+      return deferred
     }
 
     private fun generateFactoryClass(
       clazz: KSClassDeclaration,
       constructor: KSFunctionDeclaration,
-    ): FileSpec {
+    ): FileSpec? {
       val typeParameterResolver = clazz.typeParameters.toTypeParameterResolver()
       val constructorParameters = constructor.parameters
-        .mapToConstructorParameters(typeParameterResolver)
-      val memberInjectParameters = clazz.memberInjectParameters()
+        .mapToConstructorParameters(typeParameterResolver) ?: return null
+      val memberInjectParameters = clazz.memberInjectParameters() ?: return null
       val typeParameters = clazz.typeParameters
 
       val spec = generateFactoryClass(
@@ -164,7 +166,7 @@ internal object AssistedInjectCodeGen : AnvilApplicabilityChecker {
     clazz: ClassName,
     memberInjectParameters: List<MemberInjectParameter>,
     typeParameters: List<TypeVariableName>,
-    constructorParameters: List<ConstructorParameter>,
+    constructorParameters: Iterable<ConstructorParameter>,
     onError: (String) -> Nothing,
   ): FileSpec {
     val packageName = clazz.packageName

--- a/compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/DaggerGenerationUtils.kt
+++ b/compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/DaggerGenerationUtils.kt
@@ -11,6 +11,8 @@ import com.google.devtools.ksp.symbol.KSPropertyDeclaration
 import com.google.devtools.ksp.symbol.KSType
 import com.google.devtools.ksp.symbol.KSValueParameter
 import com.google.devtools.ksp.symbol.Visibility
+import com.squareup.anvil.compiler.addAllWithoutCopy
+import com.squareup.anvil.compiler.addWithoutCopy
 import com.squareup.anvil.compiler.assistedFqName
 import com.squareup.anvil.compiler.codegen.ksp.KspAnvilException
 import com.squareup.anvil.compiler.codegen.ksp.contextualToTypeName
@@ -118,13 +120,14 @@ private fun ParameterReference.toConstructorParameter(
 @JvmName("mapToConstructorParametersKsp")
 internal fun List<KSValueParameter>.mapToConstructorParameters(
   typeParameterResolver: TypeParameterResolver,
-): List<ConstructorParameter> {
-  return fold(listOf()) { acc, callableReference ->
-    acc + callableReference.toConstructorParameter(
-      callableReference.name!!.asString()
+): List<ConstructorParameter>? {
+  return fold(mutableListOf()) { acc, callableReference ->
+    val param = callableReference.toConstructorParameter(
+      uniqueName = callableReference.name!!.asString()
         .uniqueParameterName(acc),
-      typeParameterResolver,
-    )
+      typeParameterResolver = typeParameterResolver,
+    ) ?: return null
+    acc addWithoutCopy param
   }
 }
 
@@ -132,8 +135,9 @@ internal fun List<KSValueParameter>.mapToConstructorParameters(
 private fun KSValueParameter.toConstructorParameter(
   uniqueName: String,
   typeParameterResolver: TypeParameterResolver,
-): ConstructorParameter {
+): ConstructorParameter? {
   val type = type.resolve()
+  if (type.isError) return null
   val paramTypeName = type.contextualToTypeName(this, typeParameterResolver)
   val rawType = paramTypeName.requireRawType()
 
@@ -246,7 +250,7 @@ private fun ClassReference.declaredMemberInjectParameters(
  * The order of dependencies in `Impl_Factory`'s constructor should be:
  * Base -> Middle -> Impl
  */
-internal fun KSClassDeclaration.memberInjectParameters(): List<MemberInjectParameter> {
+internal fun KSClassDeclaration.memberInjectParameters(): List<MemberInjectParameter>? {
   // TODO can we use getAllProperties() after https://github.com/google/ksp/issues/1619?
   return sequenceOf(asType(emptyList()))
     .plus(getAllSuperTypes())
@@ -257,8 +261,9 @@ internal fun KSClassDeclaration.memberInjectParameters(): List<MemberInjectParam
       it.isInterface()
     }
     .toList()
-    .foldRight(listOf()) { classDeclaration, acc ->
-      acc + classDeclaration.declaredMemberInjectParameters(acc, this)
+    .foldRight(mutableListOf()) { classDeclaration, acc ->
+      val param = classDeclaration.declaredMemberInjectParameters(acc, this) ?: return null
+      acc addAllWithoutCopy param
     }
 }
 
@@ -270,7 +275,7 @@ internal fun KSClassDeclaration.memberInjectParameters(): List<MemberInjectParam
 private fun KSClassDeclaration.declaredMemberInjectParameters(
   superParameters: List<Parameter>,
   implementingClass: KSClassDeclaration,
-): List<MemberInjectParameter> {
+): List<MemberInjectParameter>? {
   val implementingType = implementingClass.asType(emptyList())
   return getDeclaredProperties()
     .filter {
@@ -278,14 +283,15 @@ private fun KSClassDeclaration.declaredMemberInjectParameters(
         it.setter?.isAnnotationPresent<Inject>() == true
     }
     .filter { it.getVisibility() != Visibility.PRIVATE }
-    .fold(listOf()) { acc, property ->
+    .fold(mutableListOf()) { acc, property ->
       val uniqueName = property.simpleName.asString().uniqueParameterName(superParameters, acc)
-      acc + property.toMemberInjectParameter(
+      val parameter = property.toMemberInjectParameter(
         uniqueName = uniqueName,
         declaringClass = this@declaredMemberInjectParameters,
         implementingType = implementingType,
         implementingClass = implementingClass,
-      )
+      ) ?: return null
+      acc addWithoutCopy parameter
     }
 }
 
@@ -303,7 +309,7 @@ private fun KSClassDeclaration.declaredMemberInjectParameters(
  * ```
  * Set [includeModule] to true if a Dagger module instance is part of the argument list.
  */
-internal fun List<Parameter>.asArgumentList(
+internal fun Iterable<Parameter>.asArgumentList(
   asProvider: Boolean,
   includeModule: Boolean,
 ): String {
@@ -444,7 +450,7 @@ private fun KSPropertyDeclaration.toMemberInjectParameter(
   declaringClass: KSClassDeclaration,
   implementingType: KSType,
   implementingClass: KSClassDeclaration,
-): MemberInjectParameter {
+): MemberInjectParameter? {
   if (
     !isLateInit() &&
     !isAnnotationPresent<JvmField>() &&
@@ -461,6 +467,7 @@ private fun KSPropertyDeclaration.toMemberInjectParameter(
 
   val originalName = simpleName.asString()
   val classParams = implementingClass.typeParameters.toTypeParameterResolver()
+  if (type.resolve().isError) return null
   val resolvedType = asMemberOf(implementingType)
   // TODO do we want to convert function types to lambdas?
   val propertyTypeName = resolvedType.contextualToTypeName(this, classParams)

--- a/compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/InjectConstructorFactoryCodeGen.kt
+++ b/compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/InjectConstructorFactoryCodeGen.kt
@@ -57,32 +57,34 @@ internal object InjectConstructorFactoryCodeGen : AnvilApplicabilityChecker {
     class Provider : AnvilSymbolProcessorProvider(InjectConstructorFactoryCodeGen, ::KspGenerator)
 
     override fun processChecked(resolver: Resolver): List<KSAnnotated> {
+      val deferred = mutableListOf<KSAnnotated>()
       resolver.injectConstructors()
-        .forEach { (_, constructor) ->
+        .forEach { (clazz, constructor) ->
           if (!constructor.isAnnotationPresent<Inject>()) {
             // Only generating @Inject constructors
             return@forEach
           }
 
           generateFactoryClass(constructor)
-            .writeTo(
+            ?.writeTo(
               env.codeGenerator,
               aggregating = false,
               originatingKSFiles = listOf(constructor.containingFile!!),
             )
+            ?: deferred.add(clazz)
         }
 
-      return emptyList()
+      return deferred
     }
 
     private fun generateFactoryClass(
       constructor: KSFunctionDeclaration,
-    ): FileSpec {
+    ): FileSpec? {
       val clazz = constructor.parentDeclaration as KSClassDeclaration
       val typeParameterResolver = clazz.typeParameters.toTypeParameterResolver()
       val constructorParameters =
-        constructor.parameters.mapToConstructorParameters(typeParameterResolver)
-      val memberInjectParameters = clazz.memberInjectParameters()
+        constructor.parameters.mapToConstructorParameters(typeParameterResolver) ?: return null
+      val memberInjectParameters = clazz.memberInjectParameters() ?: return null
       val typeParameters = clazz.typeParameters.map { it.toTypeVariableName(typeParameterResolver) }
 
       return generateFactoryClass(

--- a/compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/MembersInjectorCodeGen.kt
+++ b/compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/MembersInjectorCodeGen.kt
@@ -59,11 +59,24 @@ internal object MembersInjectorCodeGen : AnvilApplicabilityChecker {
     class Provider : AnvilSymbolProcessorProvider(MembersInjectorCodeGen, ::KspGenerator)
 
     override fun processChecked(resolver: Resolver): List<KSAnnotated> {
+      val deferred = mutableListOf<KSAnnotated>()
       resolver.getSymbolsWithAnnotation(injectFqName.asString())
         .mapNotNull {
           when (it) {
-            is KSPropertySetter -> SettableProperty.Setter(it)
-            is KSPropertyDeclaration -> SettableProperty.Declaration(it)
+            is KSPropertySetter -> {
+              if (it.receiver.type.resolve().isError) {
+                deferred.add(it)
+                return@mapNotNull null
+              }
+              SettableProperty.Setter(it)
+            }
+            is KSPropertyDeclaration -> {
+              if (it.type.resolve().isError) {
+                deferred.add(it)
+                return@mapNotNull null
+              }
+              SettableProperty.Declaration(it)
+            }
             else -> null
           }
         }
@@ -75,16 +88,21 @@ internal object MembersInjectorCodeGen : AnvilApplicabilityChecker {
             .map { it.toTypeVariableName() }
           val isGeneric = typeParameters.isNotEmpty()
 
+          val parameters = clazz.memberInjectParameters() ?: run {
+            deferred.add(clazz)
+            return@forEach
+          }
+
           generateMembersInjectorClass(
             origin = clazz.toClassName(),
             isGeneric = isGeneric,
             typeParameters = typeParameters,
-            parameters = clazz.memberInjectParameters(),
+            parameters = parameters,
           )
             .writeTo(env.codeGenerator, aggregating = false, listOf(clazz.containingFile!!))
         }
 
-      return emptyList()
+      return deferred
     }
 
     /**

--- a/compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/ProvidesMethodFactoryCodeGen.kt
+++ b/compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/ProvidesMethodFactoryCodeGen.kt
@@ -31,6 +31,7 @@ import com.squareup.anvil.compiler.codegen.ksp.getKSAnnotationsByType
 import com.squareup.anvil.compiler.codegen.ksp.isAnnotationPresent
 import com.squareup.anvil.compiler.codegen.ksp.isInterface
 import com.squareup.anvil.compiler.codegen.ksp.reportableReturnTypeNode
+import com.squareup.anvil.compiler.codegen.ksp.returnTypeOrNull
 import com.squareup.anvil.compiler.codegen.ksp.withCompanion
 import com.squareup.anvil.compiler.codegen.ksp.withJvmSuppressWildcardsIfNeeded
 import com.squareup.anvil.compiler.daggerModuleFqName
@@ -82,6 +83,7 @@ internal object ProvidesMethodFactoryCodeGen : AnvilApplicabilityChecker {
     class Provider : AnvilSymbolProcessorProvider(ProvidesMethodFactoryCodeGen, ::KspGenerator)
 
     override fun processChecked(resolver: Resolver): List<KSAnnotated> {
+      val deferred = mutableListOf<KSAnnotated>()
       resolver.getSymbolsWithAnnotation(daggerModuleFqName.asString())
         .filterIsInstance<KSClassDeclaration>()
         .forEach { clazz ->
@@ -91,14 +93,24 @@ internal object ProvidesMethodFactoryCodeGen : AnvilApplicabilityChecker {
               .flatMap {
                 it.getAnnotatedFunctions<Provides>()
               }
+              .filter { function ->
+                sequenceOf(function.returnType).plus(function.parameters.map { it.type })
+                  .none { type -> type?.resolve()?.isError == true }
+                  .also { hasErrors ->
+                    if (hasErrors) {
+                      deferred.add(function)
+                    }
+                  }
+              }
               .onEach { function ->
                 checkFunctionIsNotAbstract(clazz, function)
               }
               .also { functions ->
                 assertNoDuplicateFunctions(clazz, functions)
               }
-              .map { function ->
+              .mapNotNull { function ->
                 CallableReference.from(function)
+                  .also { if (it == null) deferred.add(function) }
               }
 
           val properties = classAndCompanion.flatMap { it.getDeclaredProperties() }
@@ -111,8 +123,12 @@ internal object ProvidesMethodFactoryCodeGen : AnvilApplicabilityChecker {
                 ) ||
                 property.getter?.isAnnotationPresent<Provides>() == true
             }
-            .map { property ->
-              CallableReference.from(property)
+            .mapNotNull { property ->
+              CallableReference.from(property).also {
+                if (it == null) {
+                  deferred.add(property)
+                }
+              }
             }
 
           val className = clazz.toClassName()
@@ -179,7 +195,7 @@ internal object ProvidesMethodFactoryCodeGen : AnvilApplicabilityChecker {
 
     private fun CallableReference.Companion.from(
       function: KSFunctionDeclaration,
-    ): CallableReference {
+    ): CallableReference? {
       if (function.extensionReceiver != null) {
         throw KspAnvilException(
           message = "@Provides methods can not be an extension function",
@@ -190,6 +206,7 @@ internal object ProvidesMethodFactoryCodeGen : AnvilApplicabilityChecker {
         message = "Error occurred in type resolution and could not resolve return type.",
         node = function,
       )
+      if (type.isError) return null
       val typeName = type.contextualToTypeName(
         function.reportableReturnTypeNode,
       ).withJvmSuppressWildcardsIfNeeded(function, type)
@@ -200,7 +217,7 @@ internal object ProvidesMethodFactoryCodeGen : AnvilApplicabilityChecker {
         isProperty = false,
         constructorParameters = function.parameters.mapToConstructorParameters(
           function.typeParameters.toTypeParameterResolver(),
-        ),
+        ) ?: return null,
         type = typeName,
         isNullable = type.isMarkedNullable,
         isPublishedApi = function.isAnnotationPresent<PublishedApi>(),
@@ -210,8 +227,8 @@ internal object ProvidesMethodFactoryCodeGen : AnvilApplicabilityChecker {
 
     private fun CallableReference.Companion.from(
       property: KSPropertyDeclaration,
-    ): CallableReference {
-      val type = property.type.resolve()
+    ): CallableReference? {
+      val type = property.type.resolve().takeUnless { it.isError } ?: return null
       val typeName = type.contextualToTypeName(
         property.type,
       ).withJvmSuppressWildcardsIfNeeded(property, type)

--- a/compiler/src/main/java/com/squareup/anvil/compiler/codegen/ksp/KspAnvilException.kt
+++ b/compiler/src/main/java/com/squareup/anvil/compiler/codegen/ksp/KspAnvilException.kt
@@ -1,9 +1,0 @@
-package com.squareup.anvil.compiler.codegen.ksp
-
-import com.google.devtools.ksp.symbol.KSNode
-
-internal class KspAnvilException(
-  override val message: String,
-  val node: KSNode,
-  override val cause: Throwable? = null,
-) : Exception()

--- a/compiler/src/main/java/com/squareup/anvil/compiler/codegen/ksp/KspAnvilExceptions.kt
+++ b/compiler/src/main/java/com/squareup/anvil/compiler/codegen/ksp/KspAnvilExceptions.kt
@@ -1,0 +1,18 @@
+package com.squareup.anvil.compiler.codegen.ksp
+
+import com.google.devtools.ksp.symbol.KSAnnotated
+import com.google.devtools.ksp.symbol.KSNode
+
+internal class KspAnvilException(
+  override val message: String,
+  val node: KSNode,
+  override val cause: Throwable? = null,
+) : Exception()
+
+/**
+ * In rare cases, we might encounter error types that we can't easily forward back up the call
+ * stack. With great shame, I hang my head in defeat with this type.
+ */
+internal class KspErrorTypeException(val typesToDefer: List<KSAnnotated>) : Exception() {
+  constructor(vararg types: KSAnnotated) : this(types.toList())
+}

